_posts/2025-03-26-phased-strimzi-upgrade-example.md

```diff
@@ -0,0 +1,700 @@
+
+```shell
+kubectl -n strimzi apply -k install-files
+```
+
+Note that the namespace you install the Operator Set into needs to match the one you have used in the `patch-role-references.yaml`.
+
+A full example of the Operator Set above can be seen in the [strimzi-phased-upgrade](https://github.com/tomncooper/strimzi-phased-upgrade) repository.
+That repository also contains scripting to automate the addition of new Strimzi versions and templates for the generation of patch and kustomization files. 
+
+#### The Phased Upgrade Process
+
+Now that we have a way to install multiple Strimzi versions, we can go over the process of moving Kafka clusters (and other operands) between them.
+
+Firstly, this setup requires all Kafka clusters to be labelled with the appropriate Strimzi custom resource selector.
+Lets imagine that our Kafka fleet consists of 3 clusters: A, B & C, which are all running Kafka 3.8.0 and are located in the `kafka` namespace.
+We want these clusters to be managed by the Stimzi 0.43 instance.
+So we apply the label defined in the `install-files/cluster-operator-0.43.0/patches/patch-deployment.yaml` file above: `strimzi-resource-selector=strimzi-0-43-0`.
+This can either be done in the Kafka cluster CR:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: kafka-a
+  labels:
+    strimzi-resource-selector: strimzi-0-43-0
+spec:
+  kafka:
+    version: 3.8.0
+...
+```
+
+Or by applying the label directly to an existing Kafka CR:
+
+```shell
+kubectl -n kafka label --overwrite kafka kafka-a strimzi-resource-selector=strimzi-0-43-0
+```
+
+Once we have applied this label to all the Kafka clusters in our fleet, we should see Strimzi 0.43 happily reconciling the Kafka CRs and the other Strimzi versions sitting idle.
+
+```mermaid
+flowchart TD
+	S_043(Strimzi 0.43) --> K_A(Kafka A)
+  S_043 --> K_B(Kafka B)
+  S_043 --> K_C(Kafka C)
+	S_044(Strimzi 0.44)
+```
+
+We now want to upgrade to a newer Strimzi version. 
+The Kafka C cluster is our most risk tolerant cluster, it hosts topics used by low tier data pipelines which can tolerate extended downtime, so it will go first to ensure there are no issues.
+We first check that the Kafka C cluster is error free and in the `Ready` state and then apply the Strimzi custom resource selector label for the 0.44.0 Strimzi operator:
+
+```shell
+kubectl -n kafka label --overwrite kafka kafka-c strimzi-resource-selector=strimzi-0-44-0
+```
+
+The 0.44.0 Operator instance will see the new cluster and update the image reference.
+It will then roll the cluster, one broker at a time, on to the new image.
+
+```mermaid
+flowchart TD
+  S_043(Strimzi 0.43) --> K_A(Kafka A)
+  S_043 --> K_B(Kafka B)
+	S_044(Strimzi 0.44) --> K_C(Kafka C)
+```
+
+After a burn-in period, where we are happy with the stability of the cluster, our client side canary jobs are stable and no client teams are reporting issues, we begin to move the rest of the fleet.
```

Comment:
"our client side canary jobs"  comes a bit out-of-the-blue
Will it raise any questions in the reader?
Could we keep this neutral " where the cluster remains stable and no issues are reported"

Reply:
Yes, that is a bit out of the blue. I will simplify it.